daemon/mqtt.c

```diff
@@ -0,0 +1,401 @@
+}
+
+
+void disconnect_callback(struct mosquitto *mosq, void *obj, int rc, int flags, const mosquitto_property *props)
+{
+    info("Connection to cloud failed");
+    // TODO: Keep the connection "alive" for now. The library will reconnect.
+    //mqtt_connection_initialized = 0;
+}
+
+
+void mqtt_shutdown()
+{
+    int rc;
+
+    info("MQTT Shutdown initiated");
+
+    mqtt_connection_initialized = 0;
+
+    rc = mosquitto_disconnect(mosq);
+    switch (rc) {
+        case MOSQ_ERR_SUCCESS:
+            info("MQTT disconnected from broker");
+            break;
+        default:
+            info("MQTT invalid structure");
+            break;
+    };
+    info("Thread processing shutting down");
+
+    mosquitto_destroy(mosq);
+    info("MQTT shutdown complete");
+    mosq = NULL;
+}
+
+int mqtt_init(MQTT_INIT_ACTION action)
+{
+    static  init = 0;
+    int rc;
+
+    // Check if we should do reinit
+    if (unlikely(action == MQTT_REINIT)) {
+        if (unlikely(!init))
+            return 0;
+
+        info("MQTT reinit requested");
+        mqtt_shutdown();
+        info("Cleanup mosquitto library");
+        mosquitto_lib_cleanup();
+
+    }
+
+    if (unlikely(!init)) {
+        mqtt_send_maximum  = config_get_number(CONFIG_SECTION_MQTT, "mqtt send maximum", 20);
+        mqtt_recv_maximum  = config_get_number(CONFIG_SECTION_MQTT, "mqtt receive maximum", 20);
+
+        mqtt_broker_hostname = config_get(CONFIG_SECTION_MQTT, "mqtt broker hostname",  "localhost");
+        mqtt_broker_port     = config_get_number(CONFIG_SECTION_MQTT, "mqtt broker port",  1883);
+
+        info("Maximum parallel outgoing messages %d", mqtt_send_maximum);
+        info("Maximum parallel incoming messages %d", mqtt_recv_maximum);
+
+        // This will setup the base publish topic internally
+        get_publish_base_topic(PUBLICH_TOPIC_GET);
+        init = 1;
+    }
+
+    // initialize mosquitto library
+    rc = mosquitto_lib_init();
+    if (unlikely(rc != MOSQ_ERR_SUCCESS)) {
+        error("Failed to initialize MQTT (libmosquitto library)");
+        return 1;
+    }
+
+    mosq = mosquitto_new(NULL, true, NULL);
+    if (unlikely(!mosq)) {
+        mosquitto_lib_cleanup();
+        error("MQTT new structure  -- %s", mosquitto_strerror(errno));
+        return 1;
+    }
+
+    mosquitto_connect_v5_callback_set(mosq, connect_callback);
+
+    mosquitto_disconnect_v5_callback_set(mosq, disconnect_callback);
+
+    rc = mosquitto_threaded_set(mosq,1);
+    if (unlikely(rc != MOSQ_ERR_SUCCESS))
+        error("Failed to tune the thread model for libmoquitto (%s)",mosquitto_strerror(rc));
+
+    // attempt V5 protocol for now, no doc on the return status for now
+    rc = mosquitto_int_option(mosq, MQTT_PROTOCOL_V5, 0);
+    if (unlikely(rc != MOSQ_ERR_SUCCESS))
+        error("MQTT protocol specification rc = %d (%s)", rc, mosquitto_strerror(rc));
+
+
+    // These added for the V5 -- may need to be removed if we go to 3.x
```

Comment:
Instead to remove, would not be better to comment the code?

Reply:
We will most likely go with version 5 to take advantage of  new features. It can be set to be an option and select the appropriate version as needed in the future

Reply:
The protocol design is still changing but my current estimate is that we will need v5 to do the authentication properly - we can talk about this in the groom.